From 63d5c1b2d3d6846d6b945b13edd8d9fca9046e4d Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:04 -0400
Subject: [usb] cdc-acm: fix potential urb leak and PM imbalance in write

Message-id: <1410887463-20674-473-git-send-email-dzickus@redhat.com>
Patchwork-id: 95384
O-Subject: [RHEL7 PATCH 472/591] USB: cdc-acm: fix potential urb leak and PM imbalance in write
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 183a45087d126d126e8dd1d9b2602fc129dff9ad
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:41 2014 +0200

    USB: cdc-acm: fix potential urb leak and PM imbalance in write

    Make sure to check return value of autopm get in write() in order to
    avoid urb leak and PM counter imbalance on errors.

    Fixes: 11ea859d64b6 ("USB: additional power savings for cdc-acm devices
    that support remote wakeup")

    Cc: <stable@vger.kernel.org> # v2.6.27
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 1ac6c5d..c255e77 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -668,7 +668,13 @@ static int acm_tty_write(struct tty_struct *tty,
  memcpy(wb->buf, buf, count);
  wb->len = count;
 
- usb_autopm_get_interface_async(acm->control);
+ stat = usb_autopm_get_interface_async(acm->control);
+ if (stat) {
+  wb->use = 0;
+  spin_unlock_irqrestore(&acm->write_lock, flags);
+  return stat;
+ }
+
  if (acm->susp_count) {
   usb_anchor_urb(wb->urb, &acm->delayed);
   spin_unlock_irqrestore(&acm->write_lock, flags);
-- 
1.7.1